### ISSUE TITLE
Did you ever try less epoch or less step per epoch?

### ISSUE THREAD

**Issue by mutianxu** [HuguesTHOMAS/KPConv-PyTorch#48]
Dear Thomas,

Due to 500 epochs and 500 steps/epoch in your original setting cost too much time, did you ever try less epoch or step setting but can still gain the similar performance?

Thank you so much.

**Comment by HuguesTHOMAS** [comment]
Hi @mutianxu,

The number of epoch and steps per epoch is not optimized to be the fastest. I think there is margin to gain with negligible performance loss indeed. If you want to experiment with this, here are some advice:

the network convergence speed depends on three major parameters: 
- `batch_num`: the number of element per batch (in average)
- `epoch_steps`: the number of step per epoch
- `lr_decays`: the decay applied to the learning rate at the end of each epoch

From my experience,  `batch_num` is not critical, because a larger batch num means each step will take more memory and time, but also means you will have a better gradient approximation and thus a faster convergence. Overall these two effects balance each other, so you can keep a `batch_num` around 10. 

The `epoch_steps` and the `lr_decays` have exactly the same effect on convergence. lr_decay is applied at the end of each epoch. Therefore if you multiply the decay by two, it is exactly the same as multiplying the epoch steps by 2. The only difference is that between each epoch, validation is performed, so if you dont really care about validation and want faster convergence, I would recommend increasing `epoch_steps` and reducing the decays (also reduce the number of validation steps).

That being done you will be left to play with the `lr_decays`. I have come with a handy way of defining them:
https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/94b1d87e17ec339e2aad049ce5d5d175b9cd0db1/train_SemanticKitti.py#L159

I use a dictionary so that you can apply decays only at certain epochs if you like. In my case I apply it at every epoch for and exponential decay of the learning rate. The way I define it `decay(i) = X ** (1 / N)`, means that if you start with a learning rate value `L(0) = L_0`, the value after N epoch will be `L(N) = L_0 * X` and after 2N epoch, `L(2N) = L_0 * X^2`. In my case the current formula says that the learning rate is divided by 10 every 150 epochs. 

You can try to reduce N so that the learning rate is reduced faster and thus the convergence happens in fewer epochs. You will notice that if you reduce it too much you will lose in performances. I'll let you experiment with that.

Best,
Hugues

N.B. You can also try to play with the initial learning rate value, but I found that it can strongly impact the performances.

**Comment by mutianxu** [comment]
Thanks for you thorough reply.

I'm still a little confused about:
1. Where is the "validation steps" in your code?
2. In my opinion, if I want to train faster but still maintain the performance, the first thing to do is to decrease the **max_epochs** in your code. Then for matching the decreased max_epochs, the **N** in your explanation about the lr_decays should also also be decreased for faster convergence.
3. If I increase the epoch_steps as you recommended, the training time will also increase, right? So honestly I don't think this is necessary. 

Looking forward to your opinions!
Thanks a lot!

**Comment by HuguesTHOMAS** [comment]
1. Here:
https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/94b1d87e17ec339e2aad049ce5d5d175b9cd0db1/train_S3DIS.py#L158-L159

2. You are right, but `max_epoch` should only be the last thing you care about. Although it directly controls when to stop the training, it does not have an effect on the speed at which the network converges. What I generally do is: I set it to a big value first; then during the network training I use the *plot_convergence.py* script to follow the evolution of the loss and the validation performances; when they have converged, I just stop the training. Afterwards, I can decrease the `max_epoch` to the value I found with this experiment. If you reduce the `max_epoch` first without knowing when the network converges, you risk loosing performances because the network could not have enough time to converge.

3. As I said, if you increase `epoch_steps` but also decrease the decay, you will end up with the same number of training step to reach convergence, just in fewer epochs. Therefore the convergence time will remain the same. I agree, if you keep the same  `max_epoch`, it will be a longer training, but you just have to reduce it too. In fact, in this case, you will just find a smaller number of `max_epoch` with the experiment I just explained above.

Best,
Hugues
